lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HadoopPaths.java

```diff
@@ -14,36 +14,44 @@
 package io.trino.filesystem.hdfs;
 
 import com.google.common.base.VerifyException;
+import com.google.common.collect.ImmutableSet;
 import io.trino.filesystem.Location;
 import org.apache.hadoop.fs.Path;
 
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.util.Set;
 
 public final class HadoopPaths
 {
+    private static final Set<String> S3_SCHEMES = ImmutableSet.of("s3", "s3a", "s3n");
     private HadoopPaths() {}
 
     public static Path hadoopPath(Location location)
     {
         // hack to preserve the original path for S3 if necessary
         String path = location.toString();
         Path hadoopPath = new Path(path);
-        if ("s3".equals(hadoopPath.toUri().getScheme()) && !path.equals(hadoopPath.toString())) {
-            return new Path(toPathEncodedUri(location));
+        if (hadoopPath.toUri().getScheme() != null) {
+            if (S3_SCHEMES.contains(hadoopPath.toUri().getScheme()) && !path.equals(hadoopPath.toString())) {
```

Comment:
why not something like https://github.com/trinodb/trino/pull/23081?

Reply:
Logic is exactly same as that PR. Did you mean we could have avoided introducing a variable?

Reply:
I think the question was rather related to the special fragment handling for s3a/s3n.
Probably obsolete now.